File: Smart Sweepers/src/utils/SVector2D.java
package utils;

public class SVector2D {
	private double x;
	private double y;

	public SVector2D(double x, double y) {
		super();
		this.setX(x);
		this.setY(y);
	}

	public SVector2D add(SVector2D other) {
		setX(getX() + other.getX());
		setY(getY() + other.getY());

		return this;
	}

	public SVector2D sub(SVector2D other) {
		setX(getX() - other.getX());
		setY(getY() - other.getY());

		return this;
	}

	public SVector2D subN(SVector2D other) {
		return new SVector2D(getX() - other.getX(), getY() - other.getY());
	}

	public SVector2D mul(SVector2D other) {
		setX(getX() * other.getX());
		setY(getY() * other.getY());

		return this;
	}

	public SVector2D div(SVector2D other) {
		setX(getX() / other.getX());
		setY(getY() / other.getY());

		return this;
	}

	public SVector2D mul(double scalar) {
		return new SVector2D(getX() * scalar, getY() * scalar);
	}

	public double length() {
		return Math.sqrt(getX() * getX() + getY() * getY());
	}

	public double distance(SVector2D other) {
		double x = this.x - other.x;
		double y = this.y - other.y;
		return sqrt(x * x + y * y);
	}

	public SVector2D normalize() {
		double length = length();

		return new SVector2D(getX() / length, getY() / length);
	}

	public double dotProduct(SVector2D other) {
		return getX() * other.getX() + getY() * other.getY();
	}

	public int sign(SVector2D other) {
		if (getY() * other.getX() > getX() * other.getY()) {
			return 1;
		} else {
			return -1;
		}
	}

	public double getX() {
		return x;
	}

	public void setX(double x) {
		this.x = x;
	}

	public double getY() {
		return y;
	}

	public void setY(double y) {
		this.y = y;
	}
	
	private double sqrt(double d) {
		double sqrt = Double.longBitsToDouble( ( ( Double.doubleToLongBits( d )-(1l<<52) )>>1 ) + ( 1l<<61 ) );
		
		return (sqrt + d/sqrt)/2.0;
	}
}

File: Smart Sweepers/src/SVector2D.kt
import java.lang.Double.*
import kotlin.math.*

class SVector2D(var x: Double, var y: Double) {

    operator fun plusAssign(other: SVector2D) {
        x += other.x; y += other.y
    }

    operator fun minus(other: SVector2D) =
        SVector2D(x - other.x, y - other.y)

    operator fun times(scalar: Double) =
        SVector2D(x * scalar, y * scalar)

    fun length() =
        sqrt(x * x + y * y)

    fun normalize() =
        length().let { SVector2D(x / it, y / it) }

    private fun mbSqrt(d: Double): Double {
        val sqrt = longBitsToDouble(((doubleToLongBits(d) - (1L shl 52)) shr 1) + (1L shl 61))
        return (sqrt + d / sqrt) / 2.0
    }

    fun distance(other: SVector2D): Double {
        val x = this.x - other.x
        val y = this.y - other.y
        return mbSqrt(x * x + y * y)
    }
}
